sdks/java/io/splunk/src/main/java/org/apache/beam/sdk/io/splunk/SplunkIO.java

```diff
@@ -0,0 +1,359 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.splunk;
+
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
+import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.auto.value.AutoValue;
+import java.util.concurrent.ThreadLocalRandom;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.coders.BigEndianIntegerCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.MoreObjects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An unbounded sink for Splunk's Http Event Collector (HEC).
+ *
+ * <p>For more information, see the online documentation at <a
+ * href="https://dev.splunk.com/enterprise/docs/dataapps/httpeventcollector/">Splunk HEC</a>.
+ *
+ * <h3>Writing to Splunk's HEC</h3>
+ *
+ * <p>The {@link SplunkIO} class provides a {@link PTransform} that allows writing {@link
+ * SplunkEvent} messages into a Splunk HEC end point.
+ *
+ * <p>It takes as an input a {@link PCollection PCollection&lt;SplunkEvent&gt;}, where each {@link
+ * SplunkEvent} represents an event to be published to HEC.
+ *
+ * <p>To configure a {@link SplunkIO}, you must provide at a minimum:
+ *
+ * <ul>
+ *   <li>url - HEC endpoint URL.
+ *   <li>token - HEC endpoint token.
+ * </ul>
+ *
+ * <p>The {@link SplunkIO} transform can be customized further by optionally specifying:
+ *
+ * <ul>
+ *   <li>parallelism - Number of parallel requests to the HEC.
+ *   <li>batchCount - Number of events in a single batch.
+ *   <li>disableCertificateValidation - Whether to disable ssl validation (useful for self-signed
+ *       certificates)
+ * </ul>
+ *
+ * <p>This transform will return any non-transient write failures via a {@link PCollection
+ * PCollection&lt;SplunkWriteError&gt;}, where each {@link SplunkWriteError} captures the error that
+ * occurred while attempting to write to HEC. These can be published to a dead-letter sink or
+ * reprocessed.
+ *
+ * <p>For example:
+ *
+ * <pre>{@code
+ * PCollection<SplunkEvent> events = ...;
+ *
+ * PCollection<SplunkWriteError> errors =
+ *         events.apply("WriteToSplunk",
+ *              SplunkIO.writeBuilder()
```

Comment:
In Beam, these PTransform builders are not really used as such. Do you think it makes sense to have a fluent, factory-type class that doesn't require a builder?
e.g.

```
 *                 SplunkIO.write()
 *                     .withToken(token)
 *                     .withUrl(url)
 *                     .withBatchCount(batchCount)
 *                     .withParallelism(parallelism)
 *                     .withDisableCertificateValidation(true)
```

Reply:
note the example of PubsubIO, where the builder is used internally as part of the implementation, but externally, the built transform is passed around: https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubIO.java#L931-L965

Reply:
Thanks for that input! Switched to a fluent factory pattern for the SplunkIO transform.